### ISSUE TITLE
0.14.0.dev4 new -P syntax incompatible with Nanopool

### ISSUE THREAD

**Issue by xbkingx** [ethereum-mining/ethminer#904]
Nanopool 'login' syntax uses the account's email address as its password. 
The new -P command uses '@' as the delimiter between password and pool address, leading to part of the Nanopool password being interpreted as the pool address, and an invalid endpoint error.

It should be noted that using the old -O format leads to a crash on Windows with the Nanopool syntax due to the changes in delimiters. 

Couple possible solutions (least to most work, I assume):
- Check for the last '@', not first. 
- Change the delimiters to something like @@ or @PW@.
- Use a config file with comma-newline delimiters.
- Autogenerate a config/bat file interactively using some other unlikely delimiter on first launch (like xmr-stak).
- Move pool credentials and connection options into a 'connections' config file and gpu settings to a separate 'gpu' file. Allow specifying file location on the CLI

I prefer having multiple switches rather than one long command that needs parsing, especially for troubleshooting. No mess with changing syntax, they can be separated into multiple lines for easier quick referencing, and less chance of accidentally changing the wrong value.

Or maybe I'm doing something wrong. Other info:
```
ethminer version 0.14.0.dev4+git.47a57fa (latest as of writing)
Build: windows/release/msvc
Win 10 x64 1709
```

`.\ethminer.exe -RH -U -P stratum+tcp://<WALLET>.<RIG>:<PW>@eth-us-east1.nanopool.org:9999 -v 4 --display-interval 10 --cuda-schedule auto --cuda-grid-size 4096 --cuda-block-size 256 --cuda-parallel-hash 4 --cuda-streams 4`

(I know some of these are default values. They're there for when I feel like tweaking.)

Old syntax (crashes, 'not responding')
`.\ethminer.exe -RH -U -P stratum+tcp://<WALLET>.<RIG>/<PW>@eth-us-east1.nanopool.org:9999`

Bad endpoint error:
`.\ethminer.exe -RH -U -P stratum+tcp://<WALLET>.<RIG>:user.name@domain.com@eth-us-east1.nanopool.org:9999`

No problem:
`.\ethminer.exe -RH -U -P stratum+tcp://<WALLET>.<RIG>:user.namedomain.com@eth-us-east1.nanopool.org:9999`

**Comment by satori-q3a** [comment]
You're mistaking nanopool's use of email-address as a password when it's really the account id.

I have no problems with any ethminer version using win 10 on nanopool  with:
-S eth-us-east1.nanopool.org:9999 -O etheraddress.rigname/emailaddress

**Comment by xbkingx** [comment]
> I have no problems with any ethminer version using win 10 on nanopool with:
> -S eth-us-east1.nanopool.org:9999 -O etheraddress.worker/emailaddress

The -S option is being deprecated, so I assume you're using a version before 0.14.0.dev4. I had no problems before the transition to the new -P option. Try it with the latest dev version using -P instead of -S and see if it works.

**Comment by satori-q3a** [comment]
no... I'm using 0.14 dev 4 with the -O option

**Comment by xbkingx** [comment]
from ethminer -h:
```
-O, --userpass <username.workername:password> (deprecated) Stratum login credentials
.
-P URL Specify a pool URL. Can be used multiple times. The 1st for for the primary pool, and the 2nd for the failover pool.
        URL takes the form: scheme://user[:password]@hostname:port.
.
 Example: stratum+ssl://0x012345678901234567890234567890123.miner1@ethermine.org:5555
```

**Comment by xbkingx** [comment]
Yup. Still technically dev4, which is why I posted the full version. It's the latest Appveyor artifact, but the problems related to command parsing have been in at least one other dev channel build.
https://ci.appveyor.com/project/ethereum-mining/ethminer/build/job/edfovaarot31drx6/artifacts

The next version of Ethminer changes the format for the CLI options and you won't be able to use the command you posted if the bug isn't squished.

**Comment by satori-q3a** [comment]
eliminating "-O" in favor of "-P" without warning is going break a lot of people's hearts, not to mention batch files and web sites documentations.

**Comment by xbkingx** [comment]
Couldn't agree more. They're trying to pop up a warning if they see the deprecated switches, but that warning isn't displayed until a bit further in the initialization process. Right now it either doesn't reach the deprecation warning because it halts on the connection warning, or it crashes without an error.

I assume they'll keep the old commands alive for a little while, but as of right now the whole CLI options transition is borked. I think. I'm pretty sure.

**Comment by xbkingx** [comment]
Just wanted to add that finding whether the pool uses stratum+tls or stratum+ssl or any of the other 10 options is impossible on most sites, so look out for that fun bit. 

Right now I can only get stratum+tcp  to work with Nanopool, even though I know they support TLS from the old version/instructions on the the site. Whether it's stratum, stratum1, or stratum2 and tls and/or ssl, no idea.

This seems like an unnecessarily burden on the user that should be automated at first run. Query for ssl, then tls, then tls12, and finally pop up a warning that the server doesn't support encryption and ask if they want to try 'unsecure protocols', then try tcp and http. They could also include a short list of the top 10, 20, whatever pools' info to speed things up further and reduce setup errors.

**Comment by smurfy** [comment]
Thank you for reporting this. First of the PR is a work in progress.
That said, based on the official URI Specs this would be the "correct" syntax: (in my opinion)

``.\ethminer.exe -RH -U -P stratum+tcp://<WALLET>.<RIG>:user.name@domain.com@eth-us-east1.nanopool.org:9999``

You may noticed that this is one of your examples, which seem to fail ^^

On a sidenote, or nanopool, you probaby need stratum2 (or 1) did not check that detailed. (to get hashrate reporting to work)

**Comment by xbkingx** [comment]
Correct - that statement should work, but throws an invalid endpoint error and exits.

Reported hash rate has not worked with Nanopool since at least 0.13.0. Experimented with -RH, farm-recheck early on, etc. Once saw a blip I was never able to reproduce, so I live without it. Probably something on Nanopool's side related to the way it handles multiple clients with the same name (it doesn't handle them and I'm sure I've spun up a machine or 2 and forgotten to update the bat file with the rig's name).

**Comment by satori-q3a** [comment]
Nope... the -P format for nanopool fails, since nanopool uses an email address for both notification of a rig being offline and for setting the pay out rate on the account web page as a password.

0.14.rc0 's "-P" does not like double "@" in it's parsing of the pool name, but does work without the email address.

**Comment by satori-q3a** [comment]
using the old -O -S format still works, however I'm finding that there should be an additional delay between stopping and restarting ethminer.  Immediately restarting ethminer gives a windows error, but if you wait 20 secs then it restarts without problems.

The same goes for using the new -P format (without the email address for nanopool).

**Comment by troutsoup** [comment]
yeah left mine at the deprecated settings for now and seem to be working, complains they are deprecated, but i couldn't figure out a way to make it work with the -P flag. BTW i really like the colored win 10 console, so thanks!

--stratum-protocol 1 --stratum eth-us-east1.nanopool.org:9999 --userpass 0xxxxxxxxxxxxxxzzzzzzzzzzxzxzxzxzxzx.COMPUTER/email@somewhere.org

**Comment by Angel996** [comment]
Why can't we just stick to -O -S login method? What's the point of using a long delimited string, what advantages does it provide? To my knowledge, no other miner software uses fully-qualified URLs to login to pool.

**Comment by rhlug** [comment]
@Angel996   you can, dont upgrade. 

The whole point is you can define as many -P's as you want, and they can all be different stratum protocols.  Before, your -S and -SF had to be of the same -SP.   Now they dont, and you can have more than 2.

**Comment by Angel996** [comment]
In the first place, we can have that in a separate config file -- if you mean fallback pool URLs. As for me, I don't use that, I have a separate script that changes pools, algos and miners. So, can we just leave -O -S as an alternative and not remove them from future versions?

**Comment by xbkingx** [comment]
I think the ultimate goal would be to have some form of autodiscovery, rather than forcing the user to specify protocols (with a way to override), so for now I think keeping options more granular would be the logical way to go. xmr-stak just switched to using a separate pools.txt config file and I strongly prefer their general config setup.

For those that aren't familiar with xmr-stak, they have separate files for pools (separate structures per pool that accomplish all the -P syntax does in a million times more readable format), cpu, nvidia (separate structures per Nvidia GPU with their own CUDA options), amd (ditto for AMD OpenCL), and general miner config.

Seems to me that -P could have just been "Look for the first -S, -SP, -O, etc and consider everything until the next instance of that switch (or a non-pool switch) to be the parameters for a single pool." Or it could use a numbered format, like -S(1) to specify which options go together. -P just looks messy and is prone to errors.

**Comment by rhlug** [comment]
@Angel996 
I would suggest not upgrading.  If there are speed increase due to kernel changes, you can alway backport them into the 0.14 branch to keep your config.  Its open source after all.

**Comment by Angel996** [comment]
Downloaded the latest build, there is still no reasonable solution for logging onto nanopool. The idea of having several URLs in one command line is nice, but in my opinion it's beyond miner's task. Miner task is to mine on a given URL, that's it. Let other software/script switch pools/URLs.

Guys, I really do appreciate the great work you've done, but using Urlencode to log onto Nanopool? Come on, you must be kidding... :(
